core/src/internal/event_system.c

```diff
@@ -0,0 +1,333 @@
+static void destroy_thread_row(THREAD_QUEUE_ROW* row);
+static int callback_thread_main_func(void* event_system_param);
+
+EVENTSYSTEM_HANDLE EventSystem_Init(void)
+{
+	/* Codes_SRS_EVENTSYSTEM_26_001: [ This function shall create EVENTSYSTEM_HANDLE representing the created event system. ] */
+	EVENTSYSTEM_HANDLE result = (EVENTSYSTEM_HANDLE)malloc(sizeof(struct EVENTSYSTEM_DATA));
+	if (result == NULL)
+	{
+		LogError("malloc returned NULL during event system creation");
+		/* Return as is */
+	}
+	else
+	{
+		/* NULL everything for easier free() in case of VECTOR malloc failure */
+		memset(result, 0, sizeof(struct EVENTSYSTEM_DATA));
+
+		result->internal_change_lock = Lock_Init();
+		result->thread_queue_lock = Lock_Init();
+		result->thread_queue_condition = Condition_Init();
+		/* Codes_SRS_EVENTSYSTEM_26_002: [ This function shall return NULL upon any internal error during event system creation. ] */
+		if (result->internal_change_lock == NULL || result->thread_queue_lock == NULL || result->thread_queue_condition == NULL)
+		{
+			LogError("failed to initialize event system locks or condition");
+			destroy_event_system(result);
+			result = NULL;
+		}
+		else
+		{
+			for (int i = 0; i < GATEWAY_EVENTS_COUNT; i++)
+			{
+				VECTOR_HANDLE event_callbacks = VECTOR_create(sizeof(GATEWAY_CALLBACK));
+				/* Codes_SRS_EVENTSYSTEM_26_002: [ This function shall return NULL upon any internal error during event system creation. ] */
+				if (event_callbacks == NULL)
+				{
+					LogError("failed to initialize callback vectors");
+					destroy_event_system(result);
+					result = NULL;
+					break;
+				}
+				else
+				{
+					result->event_callbacks[i] = event_callbacks;
+				}
+			}
+
+			/* callback creation might have failed */
+			if (result != NULL)
+			{
+				result->delay_when_queue_empty = 1;
+
+				result->thread_queue = list_create();
+				/* Codes_SRS_EVENTSYSTEM_26_002: [ This function shall return NULL upon any internal error during event system creation. ] */
+				if (result->thread_queue == NULL)
+				{
+					LogError("failed to create thread queue during event system init");
+					destroy_event_system(result);
+					result = NULL;
+				}
+			}
+		}
+	}
+	return result;
+}
+
+void EventSystem_AddEventCallback(EVENTSYSTEM_HANDLE event_system, GATEWAY_EVENT event_type, GATEWAY_CALLBACK callback)
+{
+	/* Codes_SRS_EVENTSYSTEM_26_012: [ This function shall log a failure and do nothing else when either `event_system` or `callback` parameters are NULL. ] */
+	if (event_system == NULL || callback == NULL)
+	{
+		LogError("null gateway events handle or callback parameters during event registration");
+	}
+	/* Codes_SRS_EVENTSYSTEM_26_011: [ This function shall register given GATEWAY_CALLBACK and call it when given GATEWAY_EVENT event happens inside of the gateway. ] */
+	else if (VECTOR_push_back(event_system->event_callbacks[event_type], &callback, 1) != 0)
```

Comment:
@avranju No need, event_callbacks are always touched on the same, main, thread. Copy of event_callbacks is used for multi-threaded purpouses